feat: Allow nested arrays and vectors in Brillig foreign calls

ggiraldez opened the pull request:
# Description

This is a mirror of noir-lang/noir#4053. From the discussion there, it was suggested that we should open the PR on this repository instead, due to the changes in the serialization format.


## Problem\*

Resolves noir-lang/noir#4052

The Brillig VM has no knowledge of the data types of array/slices elements, but it also references arrays and vectors by a pointer to a small structure in memory. This results in incorrect data passed to foreign functions when the values consist of arrays (or slices) nested inside other arrays (or slices).

## Summary\*

:warning: **THIS PR CHANGES ACIR SERIALIZATION FORMAT**

This PR adds element type descriptors to `HeapArray` and `HeapVector` Brillig structures to allow the code that gathers the foreign call parameters to correctly interpret the data to be read from the VM memory.

## Additional Context

The counterpart of is returning nested arrays/slices from foreign calls, which this PR does not implement. The existing foreign functions (and the ones required by the debugger oracle) do not require this, so it's not currently a problem.

Furthermore, implementing that feature would require some design decisions regarding how are the arrays/slices allocated in memory and by who.

To return arrays the compiler now generates code to pre-allocate the heap array, adjusting the heap pointer register and setting a register for the VM to write the results into memory. For nested arrays, it would need to pre-allocate the outer as well as the inner arrays and write the pointers to the inner arrays into memory, all before making the foreign call.

When adding slices (heap vectors) to the mix, it gets more complicated because they are dynamically sized, so it wouldn't be possible for the compiler to pre-allocate the inner arrays/slices. Also, for slices the compiler will allocate a register for the VM to write the size of the returned result, which it uses to generate code to adjust the heap pointer register after the call is complete.

It's unclear whose responsibility would be to allocate heap space when mixing arrays and slices, and the compiled code should handle it. A simpler approach may be for the compiler to generate the foreign call passing the heap pointer register and delegating all the allocations and heap pointer adjustments to the VM implementation. That would simplify generated Brillig code and move the burden on to the VM implementation.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

ggiraldez committed:
feat: Allow nested arrays and vectors in Brillig foreign calls

sirasistant commented:
I think acir.hpp in barretenberg needs to be updated with the changes in acir.cpp. Keep in mind that you'll need to replace the `throw serde::deserialization_error(...` with `throw_or_abort(...`

kevaundray commented on noir/acvm-repo/acir/src/lib.rs:
```diff
@@ -68,6 +71,7 @@ mod reflection {
         tracer.trace_simple_type::<Directive>().unwrap();
         tracer.trace_simple_type::<ValueOrArray>().unwrap();
+        tracer.trace_simple_type::<HeapValueType>().unwrap();
 
```
```suggestion
```

ggiraldez committed:
Merge remote-tracking branch 'upstream/master' into nested-arrays-to-foreign-calls

ggiraldez committed:
chore: Update acir.hpp from changes in noir subrepo

ggiraldez commented on noir/acvm-repo/brillig_vm/src/lib.rs:
```diff
@@ -223,54 +232,78 @@ impl<'a, B: BlackBoxFunctionSolver> VM<'a, B> {
-                            ForeignCallParam::Single(value) => {
-                                self.memory.write(*value_index, *value);
+                        ValueOrArray::MemoryAddress(value_index) => {
+                            assert!(*value_type == HeapValueType::Simple);
```
Ah, that's a good idea! Thanks!

ggiraldez committed:
chore: Refactor matches on values and types

Addressing the code review comments, match on value and type simultaneously
instead of just the value and then asserting on the type.

sirasistant commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/serde/acir.hpp:
```diff
@@ -1,1199 +1,1168 @@
-};
+    struct Circuit {
+        uint32_t current_witness_index;
+        std::vector<Circuit::Opcode> opcodes;
+        Circuit::ExpressionWidth expression_width;
+        std::vector<Circuit::Witness> private_parameters;
+        Circuit::PublicInputs public_parameters;
+        Circuit::PublicInputs return_values;
+        std::vector<std::tuple<Circuit::OpcodeLocation, std::string>> assert_messages;
```
I think this changes need to be removed! Also it's missing passing the formatter on the whole file

vezenovm replied:
To clarify, it should look like the previous diff without the `Circuit::` prefixes.

ggiraldez replied:
Hmm... I thought this would have the same contents as the generated `acir.cpp` with the previously mentioned replacement only. Do you have the procedure to update this file documented somewhere?

vezenovm replied:
@ggiraldez unfortunately the documentation on this process is lacking. The replacement mentioned here is the only other place. I'm going to quickly do the serialization for you and also make an issue to document this process.

vezenovm replied:
Ok the committed changes look correct. Let's see what CI says. I will document the process for any future updates

ggiraldez committed:
chore: Update acir.hpp

ggiraldez committed:
Merge remote-tracking branch 'upstream/master' into nested-arrays-to-foreign-calls

TomAFrench reviewed (changes requested):
Merging this is going to delay us being able to sync the repos until we have another release of aztec-packages. If we're not going to do an extra early release then we should hold off on this.

TomAFrench reviewed (approved):
Removing block as we're going to do an extra release.

sirasistant commented:
Merged here https://github.com/AztecProtocol/aztec-packages/pull/4478

ggiraldez commented:
> Merged here #4478

Thank you!
